relay-metrics/src/aggregator.rs

```diff
@@ -16,6 +17,275 @@ use tokio::time::Instant;
+            secondary_aggregators: secondary_aggregators
+                .into_iter()
+                .map(|c| {
+                    let Condition::Eq(Field::Namespace(namespace)) = c.condition;
+                    (
+                        namespace,
+                        aggregator::Aggregator::named(c.name, c.config.aggregator),
+                    )
+                })
+                .collect(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn default_bucket_count(&self) -> usize {
+        self.default_aggregator.bucket_count()
+    }
+
+    /// Returns an iterator of references to all the aggregators in the [`AggregatorRouter`].
+    fn aggregators_ref(&self) -> impl Iterator<Item = &aggregator::Aggregator> {
+        std::iter::once(&self.default_aggregator).chain(self.secondary_aggregators.values())
+    }
+
+    /// Returns an iterator of mutable references to all the aggregators in the [`AggregatorRouter`].
+    fn aggregators_mut(&mut self) -> impl Iterator<Item = &mut aggregator::Aggregator> {
+        std::iter::once(&mut self.default_aggregator).chain(self.secondary_aggregators.values_mut())
+    }
+
+    /// Merges the given buckets into the aggregators based on the metric namespace of each bucket.
+    pub fn merge_buckets(
+        &mut self,
+        max_total_bucket_bytes: Option<usize>,
+        project_key: ProjectKey,
+        buckets: Vec<Bucket>,
+    ) {
+        let metrics_by_namespace = buckets.into_iter().group_by(|bucket| {
+            MetricResourceIdentifier::parse(&bucket.name)
+                .map(|mri| mri.namespace)
+                .ok()
+        });
+
+        // TODO: Parse MRI only once, move validation from Aggregator here.
+        for (namespace, group) in metrics_by_namespace.into_iter() {
+            let aggregator = namespace
+                .and_then(|ns| self.secondary_aggregators.get_mut(&ns))
+                .unwrap_or(&mut self.default_aggregator);
+
+            aggregator.merge_all(
+                project_key,
+                group.collect::<Vec<Bucket>>(),
+                max_total_bucket_bytes,
+            );
+        }
+    }
+
+    /// Checks if any of the aggregators' total cost exceeds the given limit.
+    pub fn any_total_cost_exceeded(&self, max_total_bucket_bytes: Option<usize>) -> bool {
+        self.aggregators_ref()
+            .any(|agg| agg.totals_cost_exceeded(max_total_bucket_bytes))
+    }
+
+    /// Runs pop_flush_buckets on all the aggregators and returns the result alongside with the name of the aggregator.
+    pub fn pop_all_flush_buckets(
+        &mut self,
+        force: bool,
+    ) -> impl Iterator<Item = (&str, HashMap<ProjectKey, Vec<HashedBucket>>)> {
+        self.aggregators_mut().map(move |agg| {
+            let flush_buckets = agg.pop_flush_buckets(force);
+            let name = agg.name();
+
+            (name, flush_buckets)
+        })
+    }
+}
+
+impl Drop for AggregatorRouter {
```

Comment:
would have made more sense to implement on `Aggregator`, but then the `into_buckets` won't work, due to stricter borrowing rules for types that implement `drop`.